Lesson 03

GarrikDanilov opened the pull request:
Задания к уроку 3

GarrikDanilov committed:
add dz3

GarrikDanilov committed:
changed task_3_4.py, task_3_5.py

GarrikDanilov committed:
changed task_3_5.py

shishkinav commented on Danilov_Garrik_dz_3/task_3_1.py:
```diff
@@ -0,0 +1,17 @@
+
+def num_translate(value: str) -> str:
+    """Переводит числительные с английского на русский"""
+    str_out = numerals.get(value)
+    return str_out
```
действительно было сложно, не так ли? )))

shishkinav commented on Danilov_Garrik_dz_3/task_3_2.py:
```diff
@@ -0,0 +1,19 @@
+def num_translate_adv(value: str) -> str:
+    """Переводит числительные с английского на русский"""
+    str_out = numerals.get(value.lower())
+    if str_out is not None and value.istitle():
```
подумайте, почему логическое выражение можно переписать вот так
```
if str_out and value.istitle():
```

shishkinav replied:
`bool()` Вам в помощь...

shishkinav commented on Danilov_Garrik_dz_3/task_3_4.py:
```diff
@@ -0,0 +1,34 @@
+
+def thesaurus_adv(*args):
+    dict_out = {}
+    for first_name, last_name in map(lambda string: string.split(' '), args):
```
не стоит использовать для переменных имена `string`, которое используется в одном из модулей

shishkinav commented on Danilov_Garrik_dz_3/task_3_4.py:
```diff
@@ -0,0 +1,34 @@
+            else:
+                dict_out[last_name[0]].update({first_name[0]: [f'{first_name} {last_name}']})
+        else:
+            dict_out[last_name[0]] = {first_name[0]: [f'{first_name} {last_name}']}
```
можно было бы ещё посмотреть в сторону оптимизации, используя метод словаря `.setdefault()`

shishkinav commented on Danilov_Garrik_dz_3/task_3_5.py:
```diff
@@ -0,0 +1,70 @@
+        :rtype: list
+        :return: list_out[str] список шуток
+    """
+    if not count:
+        return []
```
`count` является позиционным аргументом, если для него не передать значение, то функция не запустится - будет ошибка, поэтому эта проверка лишняя

shishkinav commented on Danilov_Garrik_dz_3/task_3_5.py:
```diff
@@ -0,0 +1,70 @@
+    """
+    if not count:
+        return []
+    if not flag:
+        return get_jokes(count)
```
хорошо, что переиспользуете код, но в данном случае вы фактически прокачиваете ту фукнцию, на которую ссылаетесь, а значит в один прекрасный момент может оказаться, что ссылаться некуда ))

shishkinav replied:
Вам не кажется, что в качестве флага напрашивается `bool` тип, а не `int`?

shishkinav commented on Danilov_Garrik_dz_3/task_3_5.py:
```diff
@@ -0,0 +1,70 @@
+        return []
+    if not flag:
+        return get_jokes(count)
+    min_count = min((len(nouns), len(adverbs), len(adjectives)))
+    if min_count < count:
+        print(f'Количество шуток должно быть не более {min_count}')
+        return []
```
кажется здесь вы лишаете функцию возможности работать без требования уникальности, т.е. если мне не нужна будет уникальность и я попрошу из тех наборов, которые есть сложить мне 50 шуток, то в Вашей реализации я получу `[]`, что кажется неправильным

shishkinav replied:
лишние скобки и можно было добавить сюда `count`, чтобы определить какое максимальное количество шуток мы будем отдавать на текущих наборах
```
min_count = min(len(nouns), len(adverbs), len(adjectives), count)
```

GarrikDanilov replied:
Здравствуйте
"кажется здесь вы лишаете функцию возможности работать без требования уникальности, т.е. если мне не нужна будет уникальность и я попрошу из тех наборов, которые есть сложить мне 50 шуток, то в Вашей реализации я получу [], что кажется неправильным"

Если уникальность не требуется, функция вызывается в виде print(get_jokes_adv(любое число)) или print(get_jokes_adv(любое число, flag=0)). В таком случае выполняется участок кода:
    if not flag:
        return get_jokes(любое число)
возвращающий список шуток в количестве "любое число"

shishkinav commented on Danilov_Garrik_dz_3/task_3_5.py:
```diff
@@ -0,0 +1,70 @@
+
+    # Функция определения повторяющихся слов
+    def check_match(word: str):
+        for string in list_jokes:
+            if string.find(word) == -1:
```
не путайте себя и коллег, не нужно называть переменные `string` - существует такой модуль

GarrikDanilov committed:
fixed 'if str_out is not None' in task_3_2.py;
used .setdefault() in task_3_3.py, task_3_4.py;
fixed name 'string' in task_3_4.py, task_3_5.py
